builder/proxmox/clone/builder.go

```diff
@@ -0,0 +1,77 @@
+package proxmoxclone
+
+import (
+	proxmoxapi "github.com/Telmate/proxmox-api-go/proxmox"
+	"github.com/hashicorp/hcl/v2/hcldec"
+	proxmox "github.com/hashicorp/packer/builder/proxmox/common"
+	"github.com/hashicorp/packer/helper/communicator"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+
+	"context"
+	"fmt"
+)
+
+// The unique id for the builder
+const BuilderID = "proxmox.clone"
+
+type Builder struct {
+	config Config
+}
+
+// Builder implements packer.Builder
+var _ packer.Builder = &Builder{}
+
+func (b *Builder) ConfigSpec() hcldec.ObjectSpec { return b.config.FlatMapstructure().HCL2Spec() }
+
+func (b *Builder) Prepare(raws ...interface{}) ([]string, []string, error) {
+	return b.config.Prepare(raws...)
+}
+
+func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (packer.Artifact, error) {
+	state := new(multistep.BasicStateBag)
+	state.Put("clone-config", &b.config)
+	state.Put("comm", &b.config.Comm)
+
+	preSteps := []multistep.Step{
+		&StepSshKeyPair{
+			Debug:        b.config.PackerDebug,
+			DebugKeyPath: fmt.Sprintf("%s.pem", b.config.PackerBuildName),
+			Comm:         &b.config.Comm,
+		},
+	}
+	postSteps := []multistep.Step{}
+
+	sb := proxmox.NewSharedBuilder(BuilderID, b.config.Config, preSteps, postSteps, &cloneVMCreator{})
+	return sb.Run(ctx, ui, hook, state)
+}
+
+type cloneVMCreator struct{}
+
+func (*cloneVMCreator) Create(vmRef *proxmoxapi.VmRef, config proxmoxapi.ConfigQemu, state multistep.StateBag) error {
+	client := state.Get("proxmoxClient").(*proxmoxapi.Client)
+	c := state.Get("clone-config").(*Config)
+	comm := state.Get("comm").(*communicator.Config)
+
+	fullClone := 1
+	if c.FullClone == false {
+		fullClone = 0
+	}
```

Comment:
This sort of "tri-state booleans" (true/false/not set) are a bit tricky, and I think this probably doesn't behave like you intended. If the user didn't specify `full_clone` at all, it'll be `false`, meaning it doesn't match the documented default of `true`.
I don't _think_ `mapstructure` allows `*bool` types, sadly. The easier options would be either using a string `full`/`shallow` (with corresponding input validation), or switching the naming around to `shallow_clone`, since the default 0 would then turn into `full` as you want here.

Reply:
argh, go figure. I forgot that false is nil-ish in some languages (a common gotcha when traversing languages, unfortunately!).

I'll check this

Reply:
we have this Trilean type that can be used just like that:

https://github.com/hashicorp/packer/blob/c2d3e494b6874e54a28e609f752aacbb7200f8b3/builder/alicloud/ecs/image_config.go#L52

Reply:
Ah, that's convenient! :clap: